File: python/322.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        MAX = amount + 1;
        dp = [MAX for _ in range(amount + 1)]
        dp[0] = 0
        for i in range(1, len(dp)):
            for j in range(len(coins)):
                if coins[j] <= i:
                    dp[i] = min(dp[i], dp[i - coins[j]] + 1)
        if dp[amount] > amount:
            return -1
        return dp[amount]

File: java/322.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int MAX = amount + 1;
        List< Integer> dp = new ArrayList<>(Collections.nCopies(amount + 1, MAX));
        dp.set(0, 0);
        for (int i = 1; i <= amount; i++) {
            for (int j = 0; j < coins.length; j++) {
                if (coins[j] <= i) {
                    int temp = Math.min(dp.get(i), dp.get(i - coins[j]) + 1);
                    dp.set(i, temp);
                }
            }
        }
        return dp.get(amount) > amount ? -1 : dp.get(amount);
    }
}
